[AArch64] Report icmp as free if it can be folded into ands

AZero13 opened the pull request:
Since changing the backend to fold x >= 1 / x < 1 -> x > 0 / x <= 0 and x <= -1 / x > -1 -> x > 0 / x <= 0,  this should be reflected in the cost.

AZero13 commented:
@davemgreen Thoughts?

davemgreen reviewed (commented):
Hi - I think this sounds OK, some of the cases will be canonicalized by the mid-end to compare against 0. Out of interest do you have some motivating example you are trying to optimize, or is this a general improvement as you were looking at the backend patterns?

Can you add some extra tests for each of the cases?

SamTebbs33 commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -4346,14 +4346,23 @@ InstructionCost AArch64TTIImpl::getCmpSelInstrCost(
+    if (match(I->getOperand(1), m_Zero()))
+      return 0;
+
+    // x >= 1 / x < 1 -> x > 0 / x <= 0
```
I think you could explain that these patterns get folded into an `ands`, to show why their cost is zero.

AZero13 committed:
Pre-commit tests (NFC)

AZero13 committed:
[AArch64] Report icmp as free if it can be folded into ands

Since changing the backend to fold x >= 1 / x < 1 -> x > 0 / x <= 0 and x <= -1 / x > -1 -> x > 0 / x <= 0, this should be reflected in the cost.

SamTebbs33 reviewed (approved):
LGTM, thanks. Do let @davemgreen have another look though.

davemgreen reviewed (approved):
Sure, LGTM. Are you happy for this to be merged?

AZero13 commented:
> Sure, LGTM. Are you happy for this to be merged?

Yes
